src/extension/popup/PopupWindow.re

```diff
@@ -33,20 +85,52 @@ module PopupWindow = {
     };
     None
   };
-  let render {props, state, updater, handler} =>
+  let handleInLangChange {props, state} e => {
+    let inLanguage =
+      e |> ReactEventRe.Synthetic.target |> LocalDom.Element.value |> languageSettingOfString;
+    props.onInputChanged
+      inLang::(stringOfLanguageSetting inLanguage)
+      outLang::(stringOfLanguageSetting state.outLanguage)
+      props.inText;
+    Some {...state, inLanguage}
+  };
+  let handleOutLangChange {props, state} e => {
+    let outLanguage =
+      e |> ReactEventRe.Synthetic.target |> LocalDom.Element.value |> languageSettingOfString;
+    props.onInputChanged
```

Comment:
What's `inLangChanged` and `outLangChanged` for if they're not used here? Also, lots of duplication here as well, seems ripe for some helpers.